### PR TITLE
osd: Fix a race between ContextQueue and op_wq

### PR DESCRIPTION
**Pull request by majianpeng**
there is a race, see the following situation:

        ThreadA                             		                   ThreadB
  sdata->shard_lock.Lock();
  if (sdata->pqueue->empty() &&
     !(is_smallest_thread_index && !sdata->context_queue.empty())) {

								    void queue(list<Context *>& ls) {
								        bool empty = false;
                                                                       {
                                                                         std::scoped_lock l(q_mutex);
                                                                         if (q.empty()) {
                                                                           q.swap(ls);
                                                                           empty = true;
                                                                         } else {
                                                                           q.insert(q.end(), ls.begin(), ls.end());
                                                                         }
                                                                       }

                                                                       if (empty) {
                                                                         mutex.Lock();
                                                                         cond.Signal();
                                                                         mutex.Unlock();
                                                                       }
                                                                    }

     sdata->sdata_wait_lock.Lock();
    if (!sdata->stop_waiting) {

In fact mutex = sdata_wait_lock. Before ThreadA get
sdata_wait_lock.Lock(), The ThreadB may already Siganl. This because
check context_queue.empty() don't protect by sdata_wait_lock.

In fact, the root reason is for pqueue nad context_queue with the differnet lock to
protect.

Fixes: http://tracker.ceph.com/issues/36473
Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by majianpeng** [comment]
@xiexingguo @liewegas  please review.  BTW: why not  combine shard_lock and sdata_wait_lock to one?

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -10203,8 +10203,32 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
...
+	dout(20) << __func__ << " need return immediately" << dendl;
+	sdata->sdata_wait_lock.Unlock();
+	sdata->shard_lock.Unlock();
+	return;
+      }
+    } else {
+      sdata->sdata_wait_lock.Unlock();
+    }
+  } else if (sdata->pqueue->empty()) {
     sdata->sdata_wait_lock.Lock();
```
Comment (xiexingguo) [inline src/osd/OSD.cc]: Wouldn't it be simpler to recheck the context_queue after we reacquire the  sdata_wait_lock? If the context_queue is not empty, then we don't wait...

Comment (majianpeng) [reply src/osd/OSD.cc]: This ` else if (sdata->pqueue->empty())' for "if (is_smallest_thread_index && sdata->pqueue->empty()) {".  at this moment,  we don't get the sdata_wait_lock.

Comment (xiexingguo) [reply src/osd/OSD.cc]: I wouldn't mind the check below. It will return immediately and the thread will be re-scheduled soon anyway...

Comment (majianpeng) [reply src/osd/OSD.cc]: Yes, the original code is also. They don't do check again until can non-empty.

**Comment by xiexingguo** [comment]
> @xiexingguo . The root is reason is in osd we have shard_lock and sdata_wait_lock. Why not remove sdata_wait_lock? using shard_lock to protect contextqueue and pqueue.

Sounds like a plan. Would you mind giving it a try?

**Comment by majianpeng** [comment]
@xiexingguo #24655

**Comment by liewegas** [comment]
On Wed, 17 Oct 2018, Xie Xingguo wrote:
> > @xiexingguo . The root is reason is in osd we have shard_lock and sdata_wait_lock. Why not remove sdata_wait_lock? using shard_lock to protect contextqueue and pqueue.
> 
> Sounds like a plan. Would you mind giving it a try?

We have separate locks because there may be some latency doing the 
condition_variable signal and we don't want to do that while holding the 
(hot) shard_lock.

**Comment by xiexingguo** [comment]
@majianpeng So shall we go ahead with this fix?

**Comment by majianpeng** [comment]
@xiexingguo . Ok ,i'll close #24655.

**Comment by neha-ojha** [comment]
This PR fixes http://tracker.ceph.com/issues/36182. Yay!
http://pulpito.ceph.com/nojha-2018-10-25_18:13:05-fs:basic_workload-wip-36182-2018-10-25-distro-basic-smithi/

**Comment by liewegas** [comment]
I think the below is a bit simpler, and avoids copy/pasting the huge hunk of code that does the wait:
```
diff --git a/src/osd/OSD.cc b/src/osd/OSD.cc
index a711a5b23b..6aaac78092 100644
--- a/src/osd/OSD.cc
+++ b/src/osd/OSD.cc
@@ -10224,17 +10224,21 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
   auto& sdata = osd->shards[shard_index];
   ceph_assert(sdata);
 
-  // If all threads of shards do oncommits, there is a out-of-order problem.
-  // So we choose the thread which has the smallest thread_index(thread_index < num_shards) of shard
-  // to do oncommit callback.
+  // If all threads of shards do oncommits, there is a out-of-order
+  // problem.  So we choose the thread which has the smallest
+  // thread_index(thread_index < num_shards) of shard to do oncommit
+  // callback.
   bool is_smallest_thread_index = thread_index < osd->num_shards;
 
   // peek at spg_t
   sdata->shard_lock.Lock();
   if (sdata->pqueue->empty() &&
-     !(is_smallest_thread_index && !sdata->context_queue.empty())) {
+      (!is_smallest_thread_index || sdata->context_queue.empty()))) {
     sdata->sdata_wait_lock.Lock();
-    if (!sdata->stop_waiting) {
+    if (is_smallest_thread_index && !sdata->context_queue.empty()) {
+      // we raced with a context_queue addition, don't wait
+      sdata->sdata_wait_lock.Unlock();
+    } else if (!sdata->stop_waiting) {
       dout(20) << __func__ << " empty q, waiting" << dendl;
       osd->cct->get_heartbeat_map()->clear_timeout(hb);
       sdata->shard_lock.Unlock();
```

**Pull request closed without merging**